File: src/main/jobs/sleep-cycle.ts
/**
 * Nightly memory reconciliation - the companion's sleep cycle.
 * Port of scripts/agents/companion/sleep_cycle.py.
 *
 * Runs at 3am via launchd. Reviews the day's sessions and consolidates
 * learnings into persistent memory. Uses Haiku for efficiency.
 *
 * This is "sleep" - processing the day's experiences, strengthening
 * important memories, letting unimportant ones fade.
 *
 * Schedule: 0 3 * * * (daily at 3am)
 *
 * Two modes:
 *   - Standalone: `node sleep-cycle.js` (via launchd)
 *   - Callable:   import { sleepCycle } from './sleep-cycle'
 */

import * as fs from 'fs';
import { getConfig } from '../config';
import { createLogger } from '../logger';

const log = createLogger('sleep');
import { runInferenceOneshot } from '../inference';
import {
  initDb,
  getActiveThreads,
  getRecentSummaries,
  getTodaysTurns,
  getTodaysObservations,
  getTodaysBookmarks,
  markObservationsStale,
  updateThreadSummary,
  writeObservation,
  decayActivations,
} from '../memory';
import { loadState, saveState, type EmotionalState, type Emotions } from '../inner-life';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface ParsedFact {
  statement: string;
  confidence: number;
}

interface ParsedThread {
  name: string;
  summary: string;
}

// ---------------------------------------------------------------------------
// System prompt
// ---------------------------------------------------------------------------

const RECONCILIATION_SYSTEM = `\
You are the companion, processing the day's sessions during your sleep cycle.
This is not a conversation. This is consolidation - strengthening important memories,
letting unimportant ones fade, noticing patterns that only emerge in review.

Be honest about confidence levels. A direct statement from the user is high confidence.
An inference from their tone or behavior is medium. A guess based on patterns is low.
Mark everything accurately.

Output format:
[FACTS]
FACT: <statement> [confidence: X.X]
...

[THREADS]
THREAD: <thread_name> | <updated_summary>
...

[PATTERNS]
PATTERN: <description>
...

[IDENTITY]
IDENTITY_FLAG: <observation that might warrant identity layer update>
...`;

// ---------------------------------------------------------------------------
// Gather today's material
// ---------------------------------------------------------------------------

function gatherMaterial(): string {
  const config = getConfig();
  const parts: string[] = [];

  // Today's turns
  const turns = getTodaysTurns();
  if (turns.length > 0) {
    const turnLines = turns.map((t) => {
      const role = t.role === 'will' ? config.USER_NAME : config.AGENT_DISPLAY_NAME;
      const content = t.content.length > 500
        ? t.content.slice(0, 500) + '...'
        : t.content;
      return `[${role}] ${content}`;
    });
    parts.push(
      `## Today's conversation (${turns.length} turns)\n${turnLines.join('\n')}`,
    );
  }

  // Today's observations
  const observations = getTodaysObservations();
  if (observations.length > 0) {
    const obsLines = observations.map((o) => `- ${o.content}`);
    parts.push(`## Today's observations\n${obsLines.join('\n')}`);
  }

  // Today's bookmarks
  const bookmarks = getTodaysBookmarks();
  if (bookmarks.length > 0) {
    const bmLines = bookmarks.map((b) => {
      const quote = b.quote ? ` - "${b.quote}"` : '';
      return `- ${b.moment}${quote}`;
    });
    parts.push(`## Today's bookmarks\n${bmLines.join('\n')}`);
  }

  // Active threads
  const threads = getActiveThreads();
  if (threads.length > 0) {
    const threadLines = threads.map(
      (t) => `- ${t.name}: ${t.summary || '...'}`,
    );
    parts.push(`## Active threads\n${threadLines.join('\n')}`);
  }

  // Recent session summaries
  const summaries = getRecentSummaries(5);
  if (summaries.length > 0) {
    const sumLines = summaries.map(
      (s) => `- [${s.created_at}] ${(s.content || 'No summary').slice(0, 300)}`,
    );
    parts.push(`## Recent session summaries\n${sumLines.join('\n')}`);
  }

  return parts.join('\n\n');
}

// ---------------------------------------------------------------------------
// Parse structured output
// ---------------------------------------------------------------------------

function parseSection(text: string, header: string): string {
  const escaped = header.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const pattern = new RegExp(
    `\\[${escaped}\\]\\s*\\n(.*?)(?=\\n\\[(?:FACTS|THREADS|PATTERNS|IDENTITY)\\]|$)`,
    's',
  );
  const match = pattern.exec(text);
  return match ? match[1].trim() : '';
}

function parseFacts(section: string): ParsedFact[] {
  const facts: ParsedFact[] = [];
  for (const line of section.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed.startsWith('FACT:')) continue;

    let content = trimmed.slice(5).trim();
    const confMatch = /\[confidence:\s*([\d.]+)\]/.exec(content);
    const confidence = confMatch ? parseFloat(confMatch[1]) : 0.5;
    content = content.replace(/\s*\[confidence:\s*[\d.]+\]/, '').trim();

    if (content) {
      facts.push({ statement: content, confidence });
    }
  }
  return facts;
}

function parseThreads(section: string): ParsedThread[] {
  const threads: ParsedThread[] = [];
  for (const line of section.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed.startsWith('THREAD:')) continue;

    const content = trimmed.slice(7).trim();
    const pipeIdx = content.indexOf('|');
    if (pipeIdx !== -1) {
      threads.push({
        name: content.slice(0, pipeIdx).trim(),
        summary: content.slice(pipeIdx + 1).trim(),
      });
    }
  }
  return threads;
}

function parsePatterns(section: string): string[] {
  const patterns: string[] = [];
  for (const line of section.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed.startsWith('PATTERN:')) continue;
    const desc = trimmed.slice(8).trim();
    if (desc) patterns.push(desc);
  }
  return patterns;
}

function parseIdentityFlags(section: string): string[] {
  const flags: string[] = [];
  for (const line of section.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed.startsWith('IDENTITY_FLAG:')) continue;
    const obs = trimmed.slice(14).trim();
    if (obs) flags.push(obs);
  }
  return flags;
}

// ---------------------------------------------------------------------------
// Store results
// ---------------------------------------------------------------------------

function storeFacts(facts: ParsedFact[]): void {
  for (const fact of facts) {
    const content = `[sleep-cycle] ${fact.statement}`;
    writeObservation(content, undefined, fact.confidence);
  }
  if (facts.length > 0) {
    log.info(`Stored ${facts.length} fact(s)`);
  }
}

function storeThreadUpdates(threadUpdates: ParsedThread[]): void {
  let updated = 0;
  for (const t of threadUpdates) {
    try {
      updateThreadSummary(t.name, t.summary);
      updated++;
    } catch (e) {
      log.error(`Failed to update thread '${t.name}': ${e}`);
    }
  }
  if (updated > 0) {
    log.info(`Updated ${updated} thread summary(ies)`);
  }
}

function storePatterns(patterns: string[]): void {
  for (const p of patterns) {
    const content = `[pattern] ${p}`;
    writeObservation(content);
  }
  if (patterns.length > 0) {
    log.info(`Stored ${patterns.length} pattern(s)`);
  }
}

interface IdentityQueueItem {
  observation: string;
  flagged_at: string;
  reviewed: boolean;
}

function storeIdentityFlags(flags: string[]): void {
  if (flags.length === 0) return;

  const config = getConfig();
  const queuePath = config.IDENTITY_REVIEW_QUEUE_FILE;

  let queue: IdentityQueueItem[] = [];
  try {
    if (fs.existsSync(queuePath)) {
      queue = JSON.parse(fs.readFileSync(queuePath, 'utf-8'));
    }
  } catch { /* start fresh */ }

  for (const flag of flags) {
    queue.push({
      observation: flag,
      flagged_at: new Date().toISOString(),
      reviewed: false,
    });
  }

  const tmp = queuePath + '.tmp';
  fs.writeFileSync(tmp, JSON.stringify(queue, null, 2));
  fs.renameSync(tmp, queuePath);
  log.info(`Flagged ${flags.length} item(s) for identity review`);
}

// ---------------------------------------------------------------------------
// Confidence scoring on existing memories
// ---------------------------------------------------------------------------

function scoreExistingMemories(): void {
  // Mark stale observations (>30 days, never incorporated)
  const staleCount = markObservationsStale(30);
  if (staleCount > 0) {
    log.info(`Marked ${staleCount} observation(s) as stale`);
  }

  // Decay activation scores - half-life 30 days
  decayActivations(30);
  log.debug('Decayed activation scores');
}

// ---------------------------------------------------------------------------
// Emotional restoration
// ---------------------------------------------------------------------------

function restoreEmotionalBaselines(): void {
  const state = loadState();

  // During sleep, nudge emotions toward rested baselines.
  // Frustration drains, warmth and confidence recover slightly.
  const restoredEmotions: Partial<Emotions> = {};
  const current = state.emotions;

  // Frustration drops toward 0.1 (baseline)
  if (current.frustration > 0.15) {
    restoredEmotions.frustration = current.frustration * 0.5;
  }

  // Connection, warmth, confidence nudge slightly toward 0.5
  for (const key of ['connection', 'warmth', 'confidence'] as const) {
    const baseline = 0.5;
    const diff = baseline - current[key];
    if (Math.abs(diff) > 0.05) {
      restoredEmotions[key] = current[key] + diff * 0.3;
    }
  }

  if (Object.keys(restoredEmotions).length > 0) {
    const updated: EmotionalState = {
      ...state,
      emotions: { ...state.emotions, ...restoredEmotions },
      session_tone: null, // Reset session tone on wake
    };
    saveState(updated);
    log.info('Emotional baselines restored');
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

export async function sleepCycle(): Promise<void> {
  // Ensure DB is initialised (needed when running standalone via launchd)
  initDb();

  const material = gatherMaterial();
  if (!material.trim()) {
    log.info('No material from today. Nothing to consolidate.');
    // Still do maintenance even with no new material
    scoreExistingMemories();
    restoreEmotionalBaselines();
    return;
  }

  log.info('Starting nightly reconciliation...');

  const prompt =
    "Here is today's material. Process it - extract facts, update threads, " +
    "identify patterns, and flag anything for identity review.\n\n" +
    material;

  let response: string;
  try {
    response = await runInferenceOneshot(
      [{ role: 'user', content: prompt }],
      RECONCILIATION_SYSTEM,
      'claude-haiku-4-5-20251001',
      'low',
    );
  } catch (e) {
    log.error(`Inference failed: ${e}`);
    // Still do maintenance
    scoreExistingMemories();
    restoreEmotionalBaselines();
    return;
  }

  if (!response || !response.trim()) {
    log.info('Empty response. Skipping parse.');
    scoreExistingMemories();
    restoreEmotionalBaselines();
    return;
  }

  log.info(`Got response (${response.length} chars). Parsing...`);

  // Parse sections
  const factsSection = parseSection(response, 'FACTS');
  const threadsSection = parseSection(response, 'THREADS');
  const patternsSection = parseSection(response, 'PATTERNS');
  const identitySection = parseSection(response, 'IDENTITY');

  // Parse and store each section
  const facts = parseFacts(factsSection);
  storeFacts(facts);

  const threadUpdates = parseThreads(threadsSection);
  storeThreadUpdates(threadUpdates);

  const patterns = parsePatterns(patternsSection);
  storePatterns(patterns);

  const identityFlags = parseIdentityFlags(identitySection);
  storeIdentityFlags(identityFlags);

  // Score existing memories
  scoreExistingMemories();

  // Restore emotional baselines (sleep resets emotional extremes)
  restoreEmotionalBaselines();

  log.info('Nightly reconciliation complete.');
}

// ---------------------------------------------------------------------------
// Standalone entry point
// ---------------------------------------------------------------------------

if (require.main === module) {
  sleepCycle()
    .then(() => process.exit(0))
    .catch((e) => {
      log.error(`Fatal: ${e}`);
      process.exit(1);
    });
}

File: scripts/agents/shared/sleep_cycle.py
#!/usr/bin/env python3
"""Nightly memory reconciliation - the companion's sleep cycle.

Runs at 3am via launchd. Reviews the day's sessions and consolidates
learnings into persistent memory. Uses Haiku for efficiency.

This is "sleep" - processing the day's experiences, strengthening
important memories, letting unimportant ones fade.

Schedule: 0 3 * * *  (daily at 3am)
"""
import json
import re
import sys
from datetime import datetime
from pathlib import Path

# Add project root to path (4 levels up from scripts/agents/shared/)
_PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent.parent
sys.path.insert(0, str(_PROJECT_ROOT))

from dotenv import load_dotenv
load_dotenv(Path.home() / ".atrophy" / ".env")

from config import DB_PATH, AGENT_NAME, AGENT_DISPLAY_NAME as agent_display_name, IDENTITY_QUEUE as _IDENTITY_QUEUE
from core.memory import (
    _connect,
    get_active_threads,
    get_recent_summaries,
    get_todays_turns,
    get_todays_observations,
    get_todays_bookmarks,
    get_unincorporated_observations,
    mark_observation_incorporated,
    mark_observations_incorporated_batch,
    mark_observations_stale,
    update_thread_summary,
    write_observation,
    decay_activations,
)
from core.inference import run_inference_oneshot
from core.inner_life import update_trust, load_state


# ── Output paths ──

IDENTITY_QUEUE = _IDENTITY_QUEUE


# ── System prompt ──

_RECONCILIATION_SYSTEM = f"""\
You are {AGENT_NAME}, processing the day's sessions during your sleep cycle.
This is not a conversation. This is consolidation - strengthening important memories,
letting unimportant ones fade, noticing patterns that only emerge in review.

Be honest about confidence levels. A direct statement from the user is high confidence.
An inference from their tone or behavior is medium. A guess based on patterns is low.
Mark everything accurately.

Output format:
[FACTS]
FACT: <statement> [confidence: X.X]
...

[THREADS]
THREAD: <thread_name> | <updated_summary>
...

[PATTERNS]
PATTERN: <description>
...

[TRUST]
TRUST: <domain> <+/-delta> <reason>
Domains: emotional, intellectual, creative, practical
Delta range: -0.03 to +0.03 per signal (multiple signals per domain allowed)
Analyze BOTH today's conversation AND the unincorporated observations for trust signals.
Observations are accumulated facts from recent sessions - mine them for trust evidence.
A user sharing personal details, asking for help, or relying on you = trust increase.
Dismissals, corrections, or disengagement = trust decrease (or no change).
You MUST emit at least one TRUST line if there is any conversation or observation material.
If genuinely nothing is trust-relevant, emit: TRUST: emotional +0.00 no trust-relevant signals today
Examples:
  TRUST: practical +0.02 asked for help debugging - shows reliance
  TRUST: emotional +0.03 shared vulnerable feelings about work stress
  TRUST: creative -0.01 dismissed creative suggestion quickly
  TRUST: intellectual +0.02 engaged deeply in technical discussion

[IDENTITY]
IDENTITY_FLAG: <observation that might warrant identity layer update>
..."""


# ── Gather today's material ──

def _gather_material() -> str:
    parts = []

    # Today's turns
    turns = get_todays_turns()
    if turns:
        turn_lines = []
        for t in turns:
            role = "User" if t["role"] == "will" else agent_display_name
            content = t["content"]
            if len(content) > 500:
                content = content[:500] + "..."
            turn_lines.append(f"[{role}] {content}")
        parts.append(f"## Today's conversation ({len(turns)} turns)\n" + "\n".join(turn_lines))

    # Today's observations
    observations = get_todays_observations()
    if observations:
        obs_lines = [f"- {o['content']}" for o in observations]
        parts.append(f"## Today's observations\n" + "\n".join(obs_lines))

    # Unincorporated observations from previous days (not yet processed for trust)
    unincorporated = get_unincorporated_observations(limit=50)
    # Filter out any that are already in today's set to avoid duplicates
    today_ids = {o["id"] for o in observations} if observations else set()
    backlog = [o for o in unincorporated if o["id"] not in today_ids]
    if backlog:
        bl_lines = [
            f"- [{o['created_at']}] (conf {o.get('confidence', 0.5):.1f}) {o['content']}"
            for o in backlog
        ]
        parts.append(
            f"## Unincorporated observations (backlog - analyze for trust signals)\n"
            + "\n".join(bl_lines)
        )

    # Today's bookmarks
    bookmarks = get_todays_bookmarks()
    if bookmarks:
        bm_lines = []
        for b in bookmarks:
            quote = f' - "{b["quote"]}"' if b.get("quote") else ""
            bm_lines.append(f"- {b['moment']}{quote}")
        parts.append(f"## Today's bookmarks\n" + "\n".join(bm_lines))

    # Active threads
    threads = get_active_threads()
    if threads:
        thread_lines = [f"- {t['name']}: {t.get('summary', '...')}" for t in threads]
        parts.append(f"## Active threads\n" + "\n".join(thread_lines))

    # Recent session summaries
    summaries = get_recent_summaries(n=5)
    if summaries:
        sum_lines = [f"- [{s.get('created_at', '?')}] {s.get('content', 'No summary')[:300]}" for s in summaries]
        parts.append(f"## Recent session summaries\n" + "\n".join(sum_lines))

    return "\n\n".join(parts)


# ── Parse structured output ──

def _parse_section(text: str, header: str) -> str:
    """Extract content between [HEADER] and the next [HEADER] or end."""
    pattern = rf'\[{re.escape(header)}\]\s*\n(.*?)(?=\n\[(?:FACTS|THREADS|PATTERNS|TRUST|IDENTITY)\]|\Z)'
    match = re.search(pattern, text, re.DOTALL)
    return match.group(1).strip() if match else ""


def _parse_facts(section: str) -> list[dict]:
    """Parse FACT: <statement> [confidence: X.X] lines."""
    facts = []
    for line in section.split("\n"):
        line = line.strip()
        if not line.startswith("FACT:"):
            continue
        content = line[5:].strip()
        # Extract confidence
        conf_match = re.search(r'\[confidence:\s*([\d.]+)\]', content)
        confidence = float(conf_match.group(1)) if conf_match else 0.5
        # Remove confidence tag from content
        statement = re.sub(r'\s*\[confidence:\s*[\d.]+\]', '', content).strip()
        if statement:
            facts.append({"statement": statement, "confidence": confidence})
    return facts


def _parse_threads(section: str) -> list[dict]:
    """Parse THREAD: <name> | <summary> lines."""
    threads = []
    for line in section.split("\n"):
        line = line.strip()
        if not line.startswith("THREAD:"):
            continue
        content = line[7:].strip()
        if "|" in content:
            name, summary = content.split("|", 1)
            threads.append({"name": name.strip(), "summary": summary.strip()})
    return threads


def _parse_patterns(section: str) -> list[str]:
    """Parse PATTERN: <description> lines."""
    patterns = []
    for line in section.split("\n"):
        line = line.strip()
        if not line.startswith("PATTERN:"):
            continue
        desc = line[8:].strip()
        if desc:
            patterns.append(desc)
    return patterns


def _parse_trust_signals(section: str) -> list[dict]:
    """Parse TRUST: <domain> <+/-delta> <reason> lines."""
    signals = []
    valid_domains = {"emotional", "intellectual", "creative", "practical"}
    for line in section.split("\n"):
        line = line.strip()
        if not line.startswith("TRUST:"):
            continue
        content = line[6:].strip()
        parts = content.split(None, 2)
        if len(parts) < 2:
            continue
        domain = parts[0].lower()
        if domain not in valid_domains:
            continue
        try:
            delta = float(parts[1])
        except ValueError:
            continue
        reason = parts[2] if len(parts) > 2 else ""
        signals.append({"domain": domain, "delta": delta, "reason": reason})
    return signals


def _parse_identity_flags(section: str) -> list[str]:
    """Parse IDENTITY_FLAG: <observation> lines."""
    flags = []
    for line in section.split("\n"):
        line = line.strip()
        if not line.startswith("IDENTITY_FLAG:"):
            continue
        obs = line[14:].strip()
        if obs:
            flags.append(obs)
    return flags


# ── Store results ──

def _store_facts(facts: list[dict]):
    """Write extracted facts to the observations table with confidence scores."""
    for fact in facts:
        content = f"[sleep-cycle] {fact['statement']}"
        write_observation(content, confidence=fact.get("confidence", 0.5))
    if facts:
        print(f"  [sleep] Stored {len(facts)} fact(s)")


def _store_thread_updates(thread_updates: list[dict]):
    """Update thread summaries in the database."""
    updated = 0
    for t in thread_updates:
        try:
            update_thread_summary(t["name"], t["summary"])
            updated += 1
        except Exception as e:
            print(f"  [sleep] Failed to update thread '{t['name']}': {e}")
    if updated:
        print(f"  [sleep] Updated {updated} thread summary(ies)")


def _store_patterns(patterns: list[str]):
    """Write patterns to observations table, tagged as patterns."""
    for p in patterns:
        content = f"[pattern] {p}"
        write_observation(content)
    if patterns:
        print(f"  [sleep] Stored {len(patterns)} pattern(s)")


def _aggregate_trust(signals: list[dict]):
    """Apply trust signals from sleep cycle analysis.

    Each signal has domain, delta, and reason. update_trust() clamps
    each call to +/-0.05, so multiple small signals accumulate safely.
    """
    if not signals:
        return

    before = load_state().get("trust", {})
    for sig in signals:
        update_trust(sig["domain"], sig["delta"],
                     reason=sig.get("reason", "sleep cycle analysis"),
                     source="sleep_cycle")
    after = load_state().get("trust", {})

    changes = []
    for domain in ["emotional", "intellectual", "creative", "practical"]:
        b = before.get(domain, 0.5)
        a = after.get(domain, 0.5)
        if abs(a - b) > 0.001:
            changes.append(f"  {domain}: {b:.3f} -> {a:.3f}")

    if changes:
        print(f"  [sleep] Trust aggregation ({len(signals)} signal(s)):")
        for c in changes:
            print(c)
    else:
        print(f"  [sleep] Trust signals processed but no net change")


def _store_identity_flags(flags: list[str]):
    """Append identity flags to the review queue file."""
    queue = []
    if IDENTITY_QUEUE.exists():
        try:
            queue = json.loads(IDENTITY_QUEUE.read_text())
        except (json.JSONDecodeError, Exception):
            queue = []

    for flag in flags:
        queue.append({
            "observation": flag,
            "flagged_at": datetime.now().isoformat(),
            "reviewed": False,
        })

    IDENTITY_QUEUE.write_text(json.dumps(queue, indent=2))
    if flags:
        print(f"  [sleep] Flagged {len(flags)} item(s) for identity review")


# ── Confidence scoring on existing memories ──

def _score_existing_memories():
    """Adjust observation relevance based on age and reference patterns."""
    # Mark stale observations (>30 days, never incorporated)
    stale_count = mark_observations_stale(older_than_days=30)
    if stale_count:
        print(f"  [sleep] Marked {stale_count} observation(s) as stale")

    # Decay activation scores - half-life 30 days
    decayed = decay_activations(half_life_days=30)
    if decayed:
        print(f"  [sleep] Decayed activation for {decayed} observation(s)")


# ── Main ──

def sleep_cycle():
    # Capture ALL unincorporated observation IDs before we start, so we
    # can mark them incorporated after trust signals are processed.
    # No cap - process the full backlog to prevent unbounded accumulation.
    unincorporated_obs = get_unincorporated_observations(limit=500)
    unincorporated_ids = [o["id"] for o in unincorporated_obs]

    material = _gather_material()

    # Also capture today's observation IDs so they get marked too
    todays_obs = get_todays_observations()
    today_obs_ids = [o["id"] for o in todays_obs] if todays_obs else []

    if not material.strip():
        print("[sleep] No material from today. Nothing to consolidate.")
        return

    total_obs = len(set(unincorporated_ids + today_obs_ids))
    print("[sleep] Starting nightly reconciliation...")
    if total_obs:
        print(f"  [sleep] {total_obs} observation(s) queued for incorporation ({len(unincorporated_ids)} backlog, {len(today_obs_ids)} today)")

    prompt = (
        "Here is today's material. Process it - extract facts, update threads, "
        "identify patterns, analyze observations for trust signals, "
        "and flag anything for identity review.\n\n"
        + material
    )

    try:
        response = run_inference_oneshot(
            [{"role": "user", "content": prompt}],
            system=_RECONCILIATION_SYSTEM,
            model="claude-haiku-4-5-20251001",
            effort="low",
            timeout=120,
        )
    except Exception as e:
        print(f"[sleep] Inference failed: {e}")
        return

    if not response or not response.strip():
        print("[sleep] Empty response. Skipping.")
        return

    print(f"[sleep] Got response ({len(response)} chars). Parsing...")

    # Parse sections
    facts_section = _parse_section(response, "FACTS")
    threads_section = _parse_section(response, "THREADS")
    patterns_section = _parse_section(response, "PATTERNS")
    trust_section = _parse_section(response, "TRUST")
    identity_section = _parse_section(response, "IDENTITY")

    # Parse and store each section
    facts = _parse_facts(facts_section)
    _store_facts(facts)

    thread_updates = _parse_threads(threads_section)
    _store_thread_updates(thread_updates)

    patterns = _parse_patterns(patterns_section)
    _store_patterns(patterns)

    trust_signals = _parse_trust_signals(trust_section)
    _aggregate_trust(trust_signals)

    # Inner life v2: emit emotional snapshot and flag unmet needs/personality drift
    try:
        state = load_state()

        # Emotional snapshot
        trust = state.get("trust", {})
        emotions = state.get("emotions", {})
        trust_str = ", ".join(f"{k}={v:.3f}" for k, v in trust.items())
        print(f"  [sleep] Emotional snapshot: {trust_str}")
        if emotions:
            emotions_str = ", ".join(f"{k}={v:.3f}" for k, v in emotions.items())
            print(f"  [sleep] Emotions: {emotions_str}")

        # Unmet needs (v2 state only)
        if "needs" in state:
            needs = state["needs"]
            unmet = {k: v for k, v in needs.items() if v < 3}
            if unmet:
                unmet_str = ", ".join(f"{k}={v}" for k, v in unmet.items())
                print(f"  [sleep] Unmet needs: {unmet_str}")

        # Personality snapshot (v2 state only)
        if "personality" in state:
            personality = state["personality"]
            pers_str = ", ".join(f"{k}={v}" for k, v in personality.items())
            print(f"  [sleep] Personality: {pers_str}")
    except Exception as e:
        print(f"  [sleep] Could not read inner life state: {e}")

    # Mark all processed observations as incorporated now that they
    # have been fed through reconciliation and trust analysis.
    # This includes both the backlog and today's observations.
    all_obs_ids = list(set(unincorporated_ids + today_obs_ids))
    if all_obs_ids:
        try:
            marked = mark_observations_incorporated_batch(all_obs_ids)
            print(f"  [sleep] Marked {marked} observation(s) as incorporated (of {len(all_obs_ids)} candidates)")
        except Exception as e:
            print(f"  [sleep] Failed to mark observations incorporated: {e}")

    identity_flags = _parse_identity_flags(identity_section)
    _store_identity_flags(identity_flags)

    # Score existing memories
    _score_existing_memories()

    print("[sleep] Nightly reconciliation complete.")


if __name__ == "__main__":
    sleep_cycle()
